### PR TITLE
HHH-16058 Removing Environment#getBytecodeProvider

### PR DESCRIPTION
**Pull request by Sanne**
https://hibernate.atlassian.net/browse/HHH-16058

cc/ @sebersole and @DavideD for awareness

### CODE REVIEW

**Comment by Sanne** [comment]
@yrodiere would you like to review this one?

File: hibernate-core/src/main/java/org/hibernate/jpa/internal/enhance/EnhancingClassTransformerImpl.java
Comment (yrodiere) [inline hibernate-core/src/main/java/org/hibernate/jpa/internal/enhance/EnhancingClassTransformerImpl.java:34]: Isn't that possibly introducing another problem? You won't be using the same bytecode provider as Hibernate ORM...

But I must admit I don't know what this class is used for anyway :/

File: hibernate-core/src/main/java/org/hibernate/jpa/internal/enhance/EnhancingClassTransformerImpl.java
Comment (yrodiere) [inline hibernate-core/src/main/java/org/hibernate/jpa/internal/enhance/EnhancingClassTransformerImpl.java:31]: There is something weird going on.

This constructor is only ever called from here:

https://github.com/hibernate/hibernate-orm/blob/a986a3806a2756b73845a2e9942b5ea86b0893fd/hibernate-core/src/main/java/org/hibernate/jpa/boot/internal/PersistenceUnitInfoDescriptor.java#L111-L114

But then all implementations of `PersistenceUnitInfo#addTransformer` are no-ops...

This component is probably not being tested. I'd suggest tring your patch in WildFly before merging?

File: hibernate-core/src/main/java/org/hibernate/proxy/AbstractLazyInitializer.java
Comment (yrodiere) [inline hibernate-core/src/main/java/org/hibernate/proxy/AbstractLazyInitializer.java:280]: Looks like you've changed the purpose of this method; maybe it should be renamed?

I don't know if this class is API or "looks like API but actually is not", though...

File: hibernate-core/src/main/java/org/hibernate/proxy/pojo/bytebuddy/SerializableProxy.java
Comment (yrodiere) [inline hibernate-core/src/main/java/org/hibernate/proxy/pojo/bytebuddy/SerializableProxy.java:121]: We won't be able to deserialize entities serialized before this patch, since the uuid will be missing.

Not sure what the backwards compatibility expectations of serialization are, though.

File: hibernate-core/src/main/java/org/hibernate/proxy/pojo/bytebuddy/SerializableProxy.java
Comment (yrodiere) [inline hibernate-core/src/main/java/org/hibernate/proxy/pojo/bytebuddy/SerializableProxy.java:137]: I know that's a pre-existing message, but don't we generally use the format "<Problem>: <reason>"?

```suggestion
			throw new IllegalStateException( "Unable to deserialize a SerializableProxy proxy: the bytecode provider is not ByteBuddy." );
```

File: hibernate-testing/src/main/java/org/hibernate/testing/bytecode/enhancement/BytecodeEnhancerRunner.java
Comment (yrodiere) [inline hibernate-testing/src/main/java/org/hibernate/testing/bytecode/enhancement/BytecodeEnhancerRunner.java:187]: You usually reset caches explicitly at the end, even if the bytecode provider is going to get garbage collected, so I suppose there's a risk of memory leak of classloader pollution or something...

But here you're not doing it? Shouldn't you keep the bytecode provider around to reset caches after the runner finishes "running"?

**Review by yrodiere — COMMENTED** [review]
Well that does looks dangerous, but I guess we like to live dangerously :) A few comments below.

By the way, does this also solve https://hibernate.atlassian.net/browse/HHH-12193 ?

**Comment by Sanne** [comment]
> By the way, does this also solve https://hibernate.atlassian.net/browse/HHH-12193 ?

Right, thanks for finding that one.

File: hibernate-testing/src/main/java/org/hibernate/testing/bytecode/enhancement/BytecodeEnhancerRunner.java
Comment (Sanne) [reply hibernate-testing/src/main/java/org/hibernate/testing/bytecode/enhancement/BytecodeEnhancerRunner.java:187]: A missing resetCaches() method shouldn't leak the classloader if it's not held in a static field (like it originally was), as it's now held only from the ServiceRegistry instance; AFAIK it's just about freeing up some memory.
I've added the `resetCaches` in other places because it's a saner approach, but it shouldn't be strictly necessary - in fact it wasn't there before.

N.B. the `resetCaches` method didn't exist in the past, I've added it recently to reduce runtime consumption of a bootstrapped Quarkus application.

File: hibernate-core/src/main/java/org/hibernate/jpa/internal/enhance/EnhancingClassTransformerImpl.java
Comment (Sanne) [reply hibernate-core/src/main/java/org/hibernate/jpa/internal/enhance/EnhancingClassTransformerImpl.java:34]: This is used by JPA's "enhance at deployment time" capability, used on EE servers.

And yes I'd like to make the BytecodeProvider implementation pluggable, but it really isn't today: choices are exclusively `bytebuddy` or `none`. Clearly this use case can't use `none`.

(there are other areas too in which it's implied that it needs to be the ByteBuddy one - I'd certainly like to improve that but I'm considering that a different issue)

File: hibernate-core/src/main/java/org/hibernate/cfg/Environment.java
Comment (Sanne) [reply hibernate-core/src/main/java/org/hibernate/cfg/Environment.java:256]: ho dear, didn't notice that was public. Thanks, restoring this - and deprecating.

File: hibernate-core/src/main/java/org/hibernate/jpa/internal/enhance/EnhancingClassTransformerImpl.java
Comment (Sanne) [reply hibernate-core/src/main/java/org/hibernate/jpa/internal/enhance/EnhancingClassTransformerImpl.java:31]: Good idea, yes WildFly has a similar implementation but it's a different one.

File: hibernate-core/src/main/java/org/hibernate/proxy/pojo/bytebuddy/SerializableProxy.java
Comment (Sanne) [reply hibernate-core/src/main/java/org/hibernate/proxy/pojo/bytebuddy/SerializableProxy.java:121]: The UUID was always supposed to be there in practice - I only noticed this wasn't happening consistently because we have some proxy serialization tests which were performing this while in the odd configuration of not having set the `sessionFactoryUuid` first.
However in a "real" scenario this would have failed as the proxy can't be initialized.

**Comment by Sanne** [comment]
pushed a reviewed version

File: hibernate-core/src/main/java/org/hibernate/jpa/internal/enhance/EnhancingClassTransformerImpl.java
Comment (yrodiere) [reply hibernate-core/src/main/java/org/hibernate/jpa/internal/enhance/EnhancingClassTransformerImpl.java:34]: > And yes I'd like to make the BytecodeProvider implementation pluggable

I was thinking more that you're not using the same *instance* of `BytecodeProvider`, whereas before this patch, you were. I have no idea what kind of impact this can have, but since the bytecode provider is stateful, it might be a problem?

**Review by yrodiere — APPROVED** [review]
Thanks. LGTM as long as you're confident about the impact on `EnhancingClassTransformerImpl.java` and WildFly.
EDIT: And serialization :)

File: hibernate-core/src/main/java/org/hibernate/proxy/AbstractLazyInitializer.java
Comment (Sanne) [reply hibernate-core/src/main/java/org/hibernate/proxy/AbstractLazyInitializer.java:277]: Good catch, I've amended the javadoc to clarify.

**Comment by Sanne** [comment]
:+1:  still checking WildFly, will merge if I spot no issues there.

**Comment by Sanne** [comment]
All good on the WildFly side - if we bypass the lack of an upodated Infinispan 2LC module.

Merging

**Pull request merged**
